src/compiler/checker.ts

```diff
@@ -6746,6 +6748,19 @@ namespace ts {
                 getPropertiesOfObjectType(type);
         }
 
+        function isTypeInvalidDueToUnionDiscriminant(contextualType: Type, obj: ObjectLiteralExpression): boolean {
+            return obj.properties.some(property => {
+                const name = property.name && getTextOfPropertyName(property.name);
+                const expected = name === undefined ? undefined : getTypeOfPropertyOfType(contextualType, name);
+                if (expected && typeIsLiteralType(expected)) {
+                    const actual = getTypeOfNode(property);
+                    // Apparently two literal types for the same literal are still not equal.
+                    return !!actual && (!typeIsLiteralType(actual) || actual.value !== expected.value);

```

Comment:
Alternatively, `findMatchingDiscriminantType` might be useful if you move _all_ the filtering logic into the checker. Although it may be less flexible than you want, since it only selects exactly one type and not many (although, if you're trying to match normal discrimination behavior, that's where it is).